test/bundler-integration/module/MultichainValidator.Module.specs.ts

```diff
@@ -0,0 +1,303 @@
+          deploymentData,
+        ]),
+        callData: batchUserOpCallData,
+        preVerificationGas: 55000,
+      },
+      smartAccountOwner,
+      entryPoint,
+      "nonce",
+    );
+    
+    // =============== make a multichain signature for a userOp ===============
+    
+    const validUntil = 0; //unlimited
+    const validAfter = 0;
+
+    const leaf1 = '0xb0bb0b'; //some random hash
+    const leaf2 = hexConcat([
+                    hexZeroPad(ethers.utils.hexlify(validUntil),6),
+                    hexZeroPad(ethers.utils.hexlify(validAfter),6),
+                    hexZeroPad(await entryPoint.getUserOpHash(deploymentUserOp),32),
+                  ]);
+    const leaf3 = '0xdecafdecaf';
+    const leaf4 = '0xa11cea11ce';
+
+    // prepare the merkle tree containing the leaves with chainId info
+    const leaves = [leaf1, leaf2, leaf3, leaf4].map(x => ethers.utils.keccak256(x));
+
+    const chainMerkleTree = new MerkleTree(
+      leaves,
+      keccak256,
+      { sortPairs: true }
+    );
+    const merkleProof = chainMerkleTree.getHexProof(leaves[1]);
+
+    const multichainSignature = await smartAccountOwner.signMessage(ethers.utils.arrayify(chainMerkleTree.getHexRoot()));
+
+    const moduleSignature = defaultAbiCoder.encode(
+      ["uint48", "uint48", "bytes32", "bytes32[]", "bytes"],
+      [
+        validUntil,
+        validAfter,
+        chainMerkleTree.getHexRoot(),
+        merkleProof,
+        multichainSignature,
+      ]
+    );
+
+    // add validator module address to the signature
+    const signatureWithModuleAddress = defaultAbiCoder.encode(
+      ["bytes", "address"],
+      [moduleSignature, multichainECDSAValidator.address]
+    );
+
+    console.log("expectedSmartAccountAddress: ", expectedSmartAccountAddress);
+    console.log("multiChainECDSAValidator: ", multichainECDSAValidator.address);
+    
+    // =================== put signature into userOp and execute ===================
+    deploymentUserOp.signature = signatureWithModuleAddress;
+    await environment.sendUserOperation(
+      deploymentUserOp,
+      entryPoint.address
+    ); 
+
+    // =================== connect SA and return everything ====================
+    const userSA = await ethers.getContractAt(
+      "SmartAccount",
+      expectedSmartAccountAddress
+    );
+
+    return {
+      entryPoint: entryPoint,
+      smartAccountImplementation: await getSmartAccountImplementation(),
+      smartAccountFactory: smartAccountFactory,
+      mockToken: mockToken,
+      ecdsaModule: ecdsaModule,
+      userSA: userSA,
+      verifyingPaymaster: await getVerifyingPaymaster(deployer, verifiedSigner),
+      multichainECDSAValidator: multichainECDSAValidator,
+      sessionKeyManager: sessionKeyManager,
+      sessionKeyMerkleTree: sessionKeyMerkleTree,
+    };
+  });
+
+  describe ("Multichain userOp validation", async () => {
+
+    it ("should process a userOp with a multichain signature", async () => {
```

Comment:
1) _add more test cases where you already have a merkle root (not generating for the first time) and you use cached leaf info to just update the tree_

I do not think it is need for the case when the tree is being build from userOps and not uploaded on-chain, but provided along with a userOp and signed. 
I do not think that dApps will store userOpHashes and userOps to then rebuild tree, I think this will be built, signed, and submitted on the go. Please refer to [this diagram](https://www.notion.so/biconomy/Multichain-Validator-module-Blog-Post-313272e5fe5d4ac88479a0ea788a5a44?pvs=4#7c099c8983b44161ab65425c8bc34a56)

2) For the session keys we have [such a test](https://github.com/bcnmy/scw-contracts/blob/adbffed2056806f9bc7a37b953ca663144a6160f/test/module/SessionKeyManager.Module.specs.ts#L103-L141). 

3) _how do we test multichain stuff using hardhat?_
We do not do it now.
```
 * @note Those tests do not actually process the userOp on several chains, instead they showcase
 * that one of the userOps included in the tree, which root is signed by user, can be processed
 * on the corresponding chain. Assuming, this will be valid for any userOp from the tree,
 * this approach can be considered representative enough for testing purposes.
 * The actual multichain tests based on Foundry framework can be added later
 ```
Can test it later on Foundry. For now I consider testing that any chain works is enough.

4) _can we add more comments at each step of test like it will enable session key on chain a and chain b and both will be validated by multichain validator? and what happens at each step._

added comments to the `test/module/MultichainValidator.Module.specs.ts`
For bundler integrated setupTests comments and flow are same that for regular.